deepchecks/nlp/checks/data_integrity/special_characters.py

```diff
@@ -83,116 +89,81 @@ def run_logic(self, context: Context, dataset_kind) -> CheckResult:
+        samples_per_special_char = {}
+        percent_special_chars_in_sample = {}
+
+        for idx, sample in zip(dataset.get_original_text_indexes(), dataset.text):
+            if pd.isna(sample):
+                continue
+            if len(sample) > self.max_chars_to_review_per_sample:
+                sample = random.sample(sample, self.max_chars_to_review_per_sample)
+            special_chars_in_sample = [char for char in sample if char in self.special_characters_deny_list]
+            percent_special_chars_in_sample[idx] = len(special_chars_in_sample) / len(sample)
+            for char in frozenset(special_chars_in_sample):
+                base = samples_per_special_char[char] if char in samples_per_special_char else []
+                samples_per_special_char[char] = base + [idx]
+
+        percents_arr = np.asarray(list(percent_special_chars_in_sample.values()))
+        percent_of_samples_with_special_chars = len(percents_arr[percents_arr > 0]) / dataset.n_samples
+        percent_special_chars_in_sample = pd.Series(percent_special_chars_in_sample).sort_values(ascending=False)
+        samples_per_special_char = dict(sorted(samples_per_special_char.items(), key=lambda x: -len(x[1])))
+        result_value = {
+            'samples_per_special_char': samples_per_special_char,
+            'percent_of_samples_with_special_chars': percent_of_samples_with_special_chars,
+            'percent_special_chars_per_sample': pd.Series(percent_special_chars_in_sample),
+        }
+
+        if context.with_display is False or len(samples_per_special_char) == 0:
             return CheckResult(value=result_value)
 
-        display_table = pd.DataFrame(
-            index=range(len(data)),
-            columns=['Special Character', '% of Samples With Character', 'Sample IDs', 'Text Example'],
-            data=[
-                [char,
-                 values['percent_of_samples'],
-                 format_list(values['samples_ids']),
-                 truncate_string(values['text_example'], self.max_text_length_for_display)]
-                for char, values in data.items()
-            ],
-        )
-        display_table = (
-            display_table.sort_values(by=['% of Samples With Character'], ascending=False)
-            .reset_index(drop=True)
-            .set_index(['Special Character'])
-        )
-        if self.n_most_common > display_table.shape[0]:
-            message = ''
-        else:
-            message = (
-                f'Showing only the top {self.n_most_common} most common special characters, '
-                'you can change it using n_most_common param.'
-            )
+        display_table = pd.DataFrame(columns=['Sample ID', '% of Special Characters',
+                                              'Special Characters', 'Text'])
+        for idx, value in percent_special_chars_in_sample[:self.max_samples_to_show].items():
+            text_sample = dataset.get_sample_at_original_index(idx)
```

Comment:
This is a bit wastful computationally. Maybe better option is to either do it all at once (guess numpy will be more efficient), or cache the location index seperatly

Reply:
It only run on the 3-10 sample that appear in the display